pydantic/main.py

```diff
@@ -266,14 +267,21 @@ class BaseModel(metaclass=ModelMetaclass):
 
     def __init__(__pydantic_self__, **data: Any) -> None:
         # Uses something other than `self` the first arg to allow "self" as a settable attribute
+        __pydantic_self__.__class__.__from_dict__(data, __pydantic_self__)
+
+    @classmethod
+    def __from_dict__(cls, data: Dict[str, Any], __pydantic_self__ = None, loc: 'LocStr' = None):
```

Comment:
I'm not sure about this, we'll need to make sure it has no performance impact.

Also I'm sure we already have a way of fixing kinks in `loc` even in sub-models. Are you sure this method is required?

I guess something like it would be required to make `loc` right in validators.

Does implementing `loc` this way allow us to simplify the type of `ErrorList` and maybe `flatten_errors`? If so it might be worth it.

Reply:
I had an idea about how to avoid this method when I remembered that you can actually call __init__ statically as well and pass the self reference. That could at least clean the code up, no statement about performance yet though. This code is not very optimized yet.

I don't know how you would fix the loc inside the validation chain as the call to validate submodels is simply a call to ModelClass(**dict) which drops the loc chain then and there.

The reason why the flatten_errors function can fix the loc is because the entire hierarchy is validated and errors exist for each nested model while the root model is passed as well. From there all the information is derived.

A validator has inherently no awareness of where the field it is validating is nested and can't derive that from the exception hierarchy as that doesn't get passed into the validator and even if it was, would be incomplete as validation is still ongoing.

As far as ErrorList is concerned, the flattening would at the very least no longer have to derive the loc at all. So that step can be removed.
It might also be possible to translate ErrorList into a simple 1D list as nesting information is no longer needed with the loc being correct already. My concern here is ordering, so that errors appear in an order that makes sense to the human reading it.

I will see what is possible here. There could be some serious performance gains on the exception path here saving on nested looping in the cases where library users need to call the errors() method (which frankly should be the majority of cases, as you need that to log errors or inform users about what they did wrong).

Reply:
agreed, good luck.